### ISSUE TITLE
Reuse code for /proc/self/maps parsing

### ISSUE THREAD

**Issue by 0x7f454c46** [checkpoint-restore/criu#690]
CRIU needs to parse `self/maps` quite often, which is not a fast interface.
Unfortunately, it's not always possible to parse it only once and reuse the cached result:
in cr-check or kerndat the usecase is: do something => check the resulting maps => process the result. So, we _actually_ need to parse it a couple of times for checks.
Though, we have like 10 functions those parse `/proc/self/maps` with slightly optimized conditions: one to find aio ring, another to find shmem's min/maj pair, another for vdso, another for stack guard page, another to detect task_size on fancy architectures,... you got the idea.
[currently adding another one]

The proposed enhancement is:
1. for CRIU actual dump/restore - parse `smaps` only once per pid and cache the result;
2. for kerndat/cr-check - it's a slow-path, so remove all the slightly optimized versions and use `parse_self_maps_lite()` or any other suitable helper to reduce the code complexity and increase code reuse.

**Comment by 0x7f454c46** [comment]
For (1) it's quite already done - but needs verification (`perf record` or whatever) that we read maps/smaps _only once_.

**Comment by 0x7f454c46** [comment]
For (2) - most of optimizations are "stop after vma found and process it".
So, we could design the helper the following way:
```
struct vm_area {
  /* addr, size, flags, file, name */
};
enum {
MAPS_PARSE_CONTINUE,
MAPS_PARSE_STOP,
};
typedef int (*parse_vma)(struct vm_area *vma, void *arg);
int parse_proc_maps(parse_vma parser, void *arg)
{
   int ret;
   FILE maps;

   fopen();
   while (true) {
        struct vm_area v;

        ret = fscanf(..., fmt, &v.addr, ... );
        if (ret == EOF || ret <= 0)
               return ...;

        ret = parser(&v, arg);
        if (ret < 0)
          goto out;
        if (ret == MAPS_PARSE_CONTINUE)
          continue;
        if (ret == MAPS_PARSE_STOP)
          break;
   }
   ret = 0;
out:
   fclose();
   return ret;
}
```
So, it would still stop on found entry and allow to reuse common parsing code.
Like for the aio:
```
int get_ring_len(struct vm_area *vma, void *arg)
{
    struct vm_area *out = arg;

    if (vma.addr != out.addr)
       return MAPS_PARSE_CONTINUE;
    if (strcmp(vma->name, "/[aio] (deleted)\n") == 0) {
       *out = *vma;
       return MAPS_PARSE_STOP;
    }
    pr_err("No AIO ring at expected location\n");
    return -1;
}
```
Or either just drop those little optimizations as it's slow-path and process the vma array/list in checks/kerndat tests.
